File: clients/rn/src/engine/room/RoomEngineCommand.ts
/**
 * Commands that can be sent to a RoomEngine.
 * TypeScript port of KMP's RoomEngineCommand.kt — discriminated union.
 */

import type { Exit, RoomObject } from '../../protocol/models';

export type RoomEngineCommand =
  | CreateRoom
  | EnterRoom
  | LeaveRoom
  | SayInRoom
  | EmoteInRoom
  | TakeObject
  | DropObject
  | UseObject
  | SelectHint
  | SetProperty;

export interface CreateRoom {
  type: 'create_room';
  name: string;
  description: string;
  zone: string;
  exits?: Exit[];
  objects?: RoomObject[];
}

export interface EnterRoom {
  type: 'enter_room';
  entityId: string;
  entityName: string;
  entityType: string;
  fromDirection: string;
}

export interface LeaveRoom {
  type: 'leave_room';
  entityId: string;
  entityName: string;
  direction: string;
}

export interface SayInRoom {
  type: 'say_in_room';
  entityId: string;
  entityName: string;
  text: string;
}

export interface TakeObject {
  type: 'take_object';
  entityId: string;
  objectName: string;
}

export interface DropObject {
  type: 'drop_object';
  entityId: string;
  objectName: string;
  objectId: string;
  description: string;
  takeable: boolean;
}

export interface UseObject {
  type: 'use_object';
  entityId: string;
  objectName: string;
  target: string | null;
}

export interface EmoteInRoom {
  type: 'emote_in_room';
  entityId: string;
  entityName: string;
  text: string;
}

export interface SelectHint {
  type: 'select_hint';
  entityId: string;
  index: number;
}

export interface SetProperty {
  type: 'set_property';
  key: string;
  value: string;
}

File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/room/RoomEngineCommand.kt
package org.wyrdsekai.app.engine.room

import org.wyrdsekai.app.protocol.Exit
import org.wyrdsekai.app.protocol.RoomObject

/**
 * Commands that can be sent to a RoomEngine.
 * Replaces Pekko's RoomCommand protocol messages.
 */
sealed class RoomEngineCommand {
    data class CreateRoom(
        val name: String,
        val description: String,
        val zone: String,
        val exits: List<Exit> = emptyList(),
        val objects: List<RoomObject> = emptyList(),
    ) : RoomEngineCommand()

    data class EnterRoom(
        val entityId: String,
        val entityName: String,
        val entityType: String,
        val fromDirection: String,
    ) : RoomEngineCommand()

    data class LeaveRoom(
        val entityId: String,
        val entityName: String,
        val direction: String,
    ) : RoomEngineCommand()

    data class SayInRoom(
        val entityId: String,
        val entityName: String,
        val text: String,
    ) : RoomEngineCommand()

    data class TakeObject(
        val entityId: String,
        val objectName: String,
    ) : RoomEngineCommand()

    data class DropObject(
        val entityId: String,
        val objectName: String,
        val objectId: String,
        val description: String,
        val takeable: Boolean,
    ) : RoomEngineCommand()

    data class UseObject(
        val entityId: String,
        val objectName: String,
        val target: String?,
    ) : RoomEngineCommand()

    data class SelectHint(
        val entityId: String,
        val index: Int,
    ) : RoomEngineCommand()

    data class EmoteInRoom(
        val entityId: String,
        val entityName: String,
        val text: String,
    ) : RoomEngineCommand()

    /** Set a room property (e.g. from onboarding flow). */
    data class SetProperty(
        val key: String,
        val value: String,
    ) : RoomEngineCommand()
}
